jupyterhub/templates/_helpers-netpol.tpl

```diff
@@ -0,0 +1,73 @@
+{{- define "jupyterhub.networkPolicy.standardEgressConfiguration" -}}
+{{- include "jupyterhub.networkPolicy.standardEgressConfiguration.withLeadingNewLine" . | trimPrefix "\n" }}
+{{- end }}
+
+{{- define "jupyterhub.networkPolicy.standardEgressConfiguration.withLeadingNewLine" -}}
+{{- $root := index . 0 }}
+{{- $netpol := index . 1 }}
+{{- if and ($netpol.egressAllowRules.dnsPortsPrivateIPs) (not $netpol.egressAllowRules.privateIPs) }}
+# Allow outbound connections to the DNS port in the private IP ranges
+- ports:
+    - protocol: UDP
+      port: 53
+    - protocol: TCP
+      port: 53
+  to:
+    - ipBlock:
+        cidr: 10.0.0.0/8
+    - ipBlock:
+        cidr: 172.16.0.0/12
+    - ipBlock:
+        cidr: 192.168.0.0/16
+{{- end }}
+
+{{- if $netpol.egressAllowRules.nonPrivateIPs }}
+# Allow outbound connections to non-private IP ranges
+{{- if $netpol.egressAllowRules.privateIPs }}
+# Allow outbound connections to private IP ranges (and their DNS ports)
```

Comment:
Are these two intentionally empty conditional blocks?

Reply:
I made them intentionally to render the comment. I felt that without them, the performance optimization to make the rules combine, made it a bit too hard to verify the logic from the rendered result based on the configuration choices.

I'm still not feeling 100% about the performance optimization of combining rules due to the complexity to read the netpol: both as a helm template and as a rendered k8s resource manifest. I'd much prefer if it was optimized effectively by a network policy controller before implemented into rules. I have asked a question related to this in a k8s sigs slack group dedicated to network policies.

I could be convinced to remove the helm template logic to render these comments, but doing that is also a nudge towards reducing the complexity fully by having clearly separate rules. Somewhat tricky judgement calls, but this can be changed without breaking anything later at least.

Reply:
Ah I don't mind having them there! Maybe add a helm comment stating why they are there?

More so than performance, I think having two different sets of rules apply to the same IP range is very confusing to reason about. You need to understand enough about netpol and iptables to *know* that the 'to' block will override the 'except' block. So the complexity gets transferred from the helm templating logic to the logic of reading and understanding what the output will do. So that is the tradeoff to be made - more conditionals in the helm logic, or conflicting rule definitions in the network policy?

Reply:
> You need to understand enough about netpol and iptables to know that the 'to' block will override the 'except' block.

Hmm..... I'd say your knowledge about the practical implementation that has been used to enforce network policies makes it more complicated. For me mainly focused on the k8s understanding, it is easier.

You have a rule saying `to`, and you can choose a target by labels or an ipBlock, the ipBlock could be a cidr range, and you are also allowed to say that it should be a cidr range with some exceptions. The destination rule is still a ipBlock which is either a full cidr range or parts of a cidr range.

I'm feeling 90% confident that a correctly implemented network policy controller/enforcer will just addatively add more persmissions based on the ipBlocks.

I had helm template comments as well, but it felt like I cluttered the logic by adding them so I deleted then :p Perhaps I can add some notes at the top instead about the implementation decisions made overall?

Reply:
> So that is the tradeoff to be made - more conditionals in the helm logic, or conflicting rule definitions in the network policy?

I'm feeling very confident that the k8s documentation is correct - network policies can't conflict - they just incrementally add permissions.

@yuvipanda you have doubt about this right? Being iptables savvy, can you verify that there can be conflicts - as I think it can't be?

Reply:
I'll try verify this now myself practically, not by inspecting the ip tables, but instead simply trying to restrict communication between pods.

Reply:
See https://github.com/jupyterhub/zero-to-jupyterhub-k8s/pull/2508#issuecomment-998753082 for a followup.